### PR TITLE
stm32: Enable DMA MUX client for nucleo_wb55rg

### PR DESCRIPTION
**Pull request by FRASTM**
Initial work to support the spi as a client of the dmamux

test on nucleo_wb55rg board 
The DT is filled with parameters to set the peripheral request on the dmamux.
The dmamux channels used for the client transfer (rx, tx) are assigned by the DT among the total nb of dmamux channels (ex.14)

Signed-off-by: Francois Ramu francois.ramu@st.com

EDIT (erwango): SPI impacts removed from this PR (moved to #23157 that has been merged separately)

### CODE REVIEW

**Comment by FRASTM** [comment]
The test application 'spi loopback' requires wired connection from MISO to MOSI of the SPI 1 on the nucleo_wb55rg. The SPI (client) transfer is using the dma and mux with request ID set to 7 and 6 as defined by the dts/arm/st/wb/stm32wb55Xg.dtsi
The test application 'spi loopback' does not consider half transfer or partial transfer when using the dma.

The SPI driver when activating the dma, is also controlling that the rx and/or tx buffers are valid before configuring the DMA Memory addresses (as a source or as a destination).

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (tbursztyka) [inline tests/drivers/spi/spi_loopback/src/spi.c]: This test seems useless. You send  buffer_rx, which name is against what the test is doing, and verify that it's content is still what you have put inside?

If the test is supposed to test null_rx, it will be like testing spi_write(). No need for any rx related spi_buf/spi_buf_set (and it will not be STM32_DMA specific at all also)

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (tbursztyka) [inline tests/drivers/spi/spi_loopback/src/spi.c]: why inverting the lengths here? And why is that stm32 dma specific as well?

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (tbursztyka) [inline tests/drivers/spi/spi_loopback/src/spi.c]: You'll need to find a way to follow the API, which is transferring 0s if tx buffer is NULL. Have a static tx buffer in your driver meant for that use-case, something like that.

**Review by tbursztyka — COMMENTED** [review]
You should not need to change the tests in spi_loopback: your driver needs to be ready to handle the various use-cases (or return -ENOTSUP if it is impossible to support some)

**Comment by FRASTM** [comment]
this PR adds the spi as a client of the dmamux for the stm32wb55 series
on top of the https://github.com/zephyrproject-rtos/zephyr/pull/22021

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (FRASTM) [reply tests/drivers/spi/spi_loopback/src/spi.c]: spi_null_rx_buf removed

**Comment by FRASTM** [comment]
rebase

**Comment by FRASTM** [comment]
changed the DTS for dma client of the SPI for nucleo_wb55
removed `compatible = "st,stm32-spi"; `  in the .overlay for this test

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (FRASTM) [reply tests/drivers/spi/spi_loopback/src/spi.c]: this is now handled in the `drivers/spi/spi_ll_stm32.c`, directly in the `spi_stm32_dma_tx_load` function : 
-->   /* if tx buff is null, then sends NOP on the line. */

File: tests/drivers/spi/spi_loopback/src/spi.c
Comment (FRASTM) [reply tests/drivers/spi/spi_loopback/src/spi.c]: spi_null_rx_buf removed

**Comment by FRASTM** [comment]
remove the macro from the dt_fixup.h

**Comment by FRASTM** [comment]
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/23157
Testcase with multiple loop on wb55rg through dmamux

File: soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series
Comment (erwango) [inline soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series:15]: Remove these lines

File: soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb55xx
Comment (erwango) [inline soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb55xx:16]: Move to soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series

**Review by erwango — CHANGES_REQUESTED** [review]
Minor comment otherwise LGTM

File: soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb55xx
Comment (FRASTM) [reply soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb55xx:14]: keeping dmamux feature for this serie only

**Comment by FRASTM** [comment]
changing syntax in Kconfig for stm32wbxx

File: soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series
Comment (FRASTM) [reply soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series]: ok

**Review by erwango — APPROVED** [review]

**Comment by FRASTM** [comment]
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/22021

**Comment by galak** [comment]
This needs rebasing to deal with merge conflict.

**Comment by FRASTM** [comment]
rebase and fix merge conflict

**Comment by carlescufi** [comment]
@galak could you merge this one if happy?

**Pull request merged**
